<div style="background-color: white; color: black; padding: 15px; border-radius: 5px; border: 1px solid #ccc;">

## 📱 Mobile Carrier Megaline - Plan Recommendation Project

**Project Description:**

We want to develop a model to analyze the subscribers' behavior who currently subscribe to legacy plans and recommend one of Megaline's newer plans: Smart or Ultra.

</div>

In [26]:
# Open and look through the data file. Path to the data file is provided below.

import pandas as pd 
df = pd.read_csv(r'C:\Users\robab\OneDrive\2025\Tripleten\AI&ML\Sprint 8\users_behavior.csv')
print(df.head())
print(df.info())
print(df.shape)

   calls  minutes  messages   mb_used  is_ultra
0     40   311.90        83  19915.42         0
1     85   516.75        56  22696.96         0
2     77   467.66        86  21060.45         0
3    106   745.53        81   8437.39         1
4     66   418.74         1  14502.75         0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   int64  
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   int64  
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 125.7 KB
None
(3214, 5)


In [27]:
#Split the source data into a training set, a validation set, and a test set.
from sklearn.model_selection import train_test_split 

# Create features (all columns except the target)
features = df.drop('is_ultra', axis=1)  # Remove target column

# Create target (the column we  want to predict)
target = df['is_ultra']

# we need THREE sets (train, validation, test), but train_test_split() only creates TWO sets at a time. So we need two separate splits.
# We Need the Two-Step Process:'
# Step 1: Split into "temp" (80%) and "test" (20%)

x_temp, x_test,y_temp, y_test = train_test_split(features, target, 
    test_size=0.2,
    random_state=42)

# Note After this split:

# X_test, y_test = 20% - LOCKED AWAY for final evaluation only
# X_temp, y_temp = 80% - AVAILABLE for training and validation
# Step 2: Split "temp" into "train" (75% of temp -> 60% of original) and "validation" (25% of temp -> 20% of original)   
x_train, x_val, y_train, y_val = train_test_split(x_temp, y_temp, 
    test_size=0.25,
    random_state=42) 
   

In [28]:
#Investigate the quality of different models by changing hyperparameters. Briefly describe the findings of the study.

# Note: Target variable: is_ultra (Ultra = 1, Smart = 0)
# Goal: Recommend which plan (Smart or Ultra)
# Task type: Classification (choosing between categories)   
from  sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (accuracy_score,precision_score,recall_score, f1_score, confusion_matrix)
# 1st_model
model = RandomForestClassifier( n_estimators =100,
                               max_depth=10,
                               min_samples_split = 5,
                               max_features = 'sqrt',
                               bootstrap = True ,
                               random_state=42)
model_trained = model.fit(x_train, y_train)
predictions = model_trained .predict(x_val)

accuracy = accuracy_score(y_val, predictions)
precision = precision_score(y_val, predictions)
recall = recall_score(y_val, predictions)
f1 = f1_score(y_val,predictions)
confusion= confusion_matrix(y_val, predictions)
# second model
model_2nd = RandomForestClassifier( n_estimators =50,
                               max_depth=5,
                               min_samples_split = 2,
                               max_features = 'sqrt',
                               bootstrap = True,
                               random_state=42)
model_trained_2nd = model_2nd.fit(x_train, y_train)
predictions_2nd = model_trained_2nd .predict(x_val)

accuracy_2nd = accuracy_score(y_val, predictions_2nd)
precision_2nd = precision_score(y_val, predictions_2nd)
recall_2nd = recall_score(y_val, predictions_2nd)
f1_2nd = f1_score(y_val,predictions_2nd)
confusion_2nd= confusion_matrix(y_val, predictions_2nd)


# 3rd_model
model_3rd = RandomForestClassifier( n_estimators =200,
                               max_depth=15,
                               min_samples_split = 20,
                               max_features = 'sqrt',
                               bootstrap = True,
                               random_state= 42)
model_trained_3rd = model_3rd.fit(x_train, y_train)
predictions_3rd = model_trained_3rd .predict(x_val)
accuracy_3rd = accuracy_score(y_val, predictions_3rd)
precision_3rd = precision_score(y_val, predictions_3rd)
recall_3rd = recall_score(y_val, predictions_3rd)
f1_3rd = f1_score(y_val,predictions_3rd)
confusion_3rd= confusion_matrix(y_val, predictions_3rd)

print(f'We do comparison between model, model_2nd and model_3rd:')
print(f"The accuracy of  model : {accuracy: .2%}, model_2nd: {accuracy_2nd:.2%}, and model_3rd : {accuracy_3rd:.2%}")
print(f"The precision of  model : {precision:.2%}, model_2nd: {precision_2nd:.2%}, and model_3rd :{precision_3rd:.2%}")
print(f"The recall of  model : {recall:.2%}, model_2nd: {recall_2nd:.2%}, and model_3rd :{recall_3rd:.2%}")
print(f"The f1 of  model : {f1:.2%}, model_2nd: {f1_2nd:.2%}, and model_3rd :{f1_3rd:.2%}")



We do comparison between model, model_2nd and model_3rd:
The accuracy of  model :  79.47%, model_2nd: 79.78%, and model_3rd : 80.87%
The precision of  model : 73.08%, model_2nd: 78.18%, and model_3rd :74.47%
The recall of  model : 49.48%, model_2nd: 44.79%, and model_3rd :54.69%
The f1 of  model : 59.01%, model_2nd: 56.95%, and model_3rd :63.06%


<div style="background-color: #e8f5e8; color: #2e7d32; padding: 20px; border-radius: 8px; border: 2px solid #4caf50; margin: 10px 0;">

## 🎯 Model Performance Analysis & Selection

### **Focus on the Primary Metric (Accuracy)**
For the Megaline project, the requirement is to achieve the highest possible accuracy with a threshold of 0.75.

<div style="color: white; background-color: #388e3c; padding: 10px; border-radius: 5px; margin: 10px 0;">

**Model Accuracy Comparison:**
• Model 1 (Accuracy): **79.47%**
• Model 2 (Accuracy): **79.78%** 
• Model 3 (Accuracy): **80.87%** ➜ **🏆 Highest Accuracy**

</div>

### **📊 Additional Performance Metrics**

**F1-Score Analysis (Balances Precision and Recall):**
<div style="color: white; background-color: #388e3c; padding: 10px; border-radius: 5px; margin: 10px 0;">

• Model 1 (F1-score): **59.01%**
• Model 2 (F1-score): **56.95%**
• Model 3 (F1-score): **63.06%** ➜ **🏆 Highest F1-score**

</div>

### **🔍 Key Hyperparameter Impact Analysis**

<div style="color: white; background-color: #388e3c; padding: 10px; border-radius: 5px; margin: 10px 0;">

•  **Impact of Trees** ($\text{n\_estimators}$): Increasing from 100 (Model 1) to 200 (Model 3) improved accuracy - more trees create a more robust ensemble

• **Impact of Depth** ($\text{max\_depth}$): Model 3's deeper trees (15) captured more complex user behavior patterns

• **Impact of Split** ($\text{min\_samples\_split}$): Model 3's higher value (20) prevented overfitting while maintaining deep learning capability

</div>

### **🎯 Final Conclusion**

<div style="background-color: #1b5e20; color: white; padding: 15px; border-radius: 5px; font-weight: bold; text-align: center; margin: 15px 0;">

**🏆 MODEL 3 IS THE OPTIMAL CHOICE** 
<br>
✅ Highest Accuracy: **80.87%** (exceeds 75% threshold)
<br>
✅ Best F1-Score: **63.06%** (superior precision-recall balance)
<br>
✅ Highest Recall: **54.69%** (best at identifying Ultra plan users)

</div>

**Model 3 achieves the highest overall correct prediction rate and demonstrates superior performance across all key metrics, making it the best configuration for the Megaline plan recommendation system.**

</div>

In [ ]:
# Check the quality of the model using the test set.

predictions = model_3rd.predict(x_test)
# Next Step: Calculate Performance Metrics
test_accuracy = accuracy_score(y_test, predictions)
test_precision = precision_score(y_test, predictions)
test_recall = recall_score(y_test, predictions)
test_f1 = f1_score(y_test, predictions)

print("=== FINAL MODEL PERFORMANCE ON TEST SET ===")
print(f"Test Accuracy: {test_accuracy:.2%}")
print(f"Test Precision: {test_precision:.2%}")
print(f"Test Recall: {test_recall:.2%}")
print(f"Test F1-Score: {test_f1:.2%}")
